File: JavaFiles/week2/Deque.java
import java.util.Iterator;
import java.util.NoSuchElementException;
import edu.princeton.cs.algs4.StdOut;

public class Deque<Item> implements Iterable<Item> {
    private Node<Item> first;
    private Node<Item> last;
    private int size;

    private static class Node<Item> {
        Item item;
        Node<Item> next;
        Node<Item> prev;

        public Node(Item item) {
            this.item = item;
            this.next = null;
            this.prev = null;
        }
    }

    public Deque() {
        first = null;
        last = null;
        size = 0;
    }

    public boolean isEmpty() {
        return first == null;
    }

    public int size() {
        return size;
    }

    public void addFirst(Item item) {
        if (item == null) throw new IllegalArgumentException("Item is null");
        size++;

        if (isEmpty()) {
            first = new Node<>(item);
            last = first;
        }
        else {
            Node<Item> oldFirst = first;
            first = new Node<>(item);
            oldFirst.prev = first;
            first.next = oldFirst;
        }
        
    }

    public void addLast(Item item) {
        if (item == null) throw new IllegalArgumentException("Item is null");
        size++;

        if (isEmpty()) {
            last = new Node<>(item);
            first = last;
        }
        else {
            Node<Item> oldLast = last;
            last = new Node<>(item);
            last.prev = oldLast;
            oldLast.next = last;
        }
        
    }

    public Item removeFirst() {
        if (isEmpty()) throw new NoSuchElementException("Deque is empty");
        size--;
        Item item = first.item;
              
        if (first.next == null) {
            first = null;
            last = null;
        }
        else {
            first = first.next;
            first.prev = null;
        }
        return item;
    }

    public Item removeLast() {
        if (isEmpty()) throw new NoSuchElementException("Deque is empty");
        size--;
        Item item = last.item;
        
        if (last.prev == null) {

            first = null;
            last = null;
        }
        else {
            last = last.prev;
            last.next = null;
        } 
        return item;
    }

    @Override
    public Iterator<Item> iterator() {
        return new DequeIterator(first);
    }

    private class DequeIterator implements Iterator<Item> {
        private Node<Item> current;

        public DequeIterator(Node<Item> first) {
            current = first;
        }

        @Override
        public boolean hasNext() {
            return current != null;
        }

        @Override
        public Item next() {
            if (current == null) throw new NoSuchElementException();
            Item item = current.item;
            current = current.next;
            return item;
        }
    }

    public static void main(String[] args) {
        Deque<Integer> deque = new Deque<>();
        deque.addFirst(1);
        deque.addLast(2);
        deque.addLast(3);
        deque.addFirst(4);
        for (int item : deque) {
            StdOut.println(item);
        }
        StdOut.println(deque.removeFirst());
        deque.removeLast();
        StdOut.println(deque.isEmpty());
        StdOut.println(deque.size());
    }
}

File: week2/deque.py
# twosided queue
class Deque:
    def __init__(self):
        self._first = None
        self._last = None
        self._size = 0

    class Node:
        def __init__(self, item):
            self.item = item
            self.next = None
            self.prev = None

    def is_empty(self):
        return self._first == None

    @property
    def size(self):
        return self._size

    def add_first(self, item):
        if item == None:
            raise Exception("Item is None")
        self._size += 1
        if self.is_empty():
            self._first = self.Node(item)
            self._last = self._first
        else:
            old_first = self._first
            self._first = self.Node(item)
            self._first.next = old_first
            old_first.prev = self._first
    
    def add_last(self, item):
        if item == None:
            raise Exception("Item is None")
        self._size += 1
        if self.is_empty():
            self._last = self.Node(item)
            self._first = self._last
        else:
            old_last = self._last
            self._last = self.Node(item)
            self._last.prev = old_last
            old_last.next = self._last

    def remove_first(self):
        if self.is_empty():
            raise Exception("Deque is empty")
        self._size -= 1
        item = self._first.item
        if self._first.next is None:
            self._first = None
            self._last = None
        else:
            self._first = self._first.next
            self._first.prev = None
        return item
        
    def remove_last(self):
        if self.is_empty():
            raise Exception("Deque is empty")
        self._size -= 1
        item = self._last.item
        if self._last.prev is None:
            self._last = None
            self._first = None
        else:
            self._last = self._last.prev
            self._last.next = None
        return item

    def __iter__(self):
        return self.DequeIterator(self)
    
    class DequeIterator:
        def __init__(self, deque):
            self._current = deque._first
        def __iter__(self):
            return self
        def has_next(self):
            return self._current.next != None
        def __next__(self):
            if self._current is None:
                raise StopIteration
            item = self._current.item
            self._current = self._current.next
            return item
